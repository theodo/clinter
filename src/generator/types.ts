import { Linter } from "eslint";
import { ProjectInfoObject } from "types";

export type ESLintGenerator = (userAnswers: ProjectInfoObject) => (config: Linter.Config) => Linter.Config;

export type ESLintDependencyGenerator = (userAnswers: ProjectInfoObject) => (deps: string[]) => string[];

export interface ESLintOverrider {
  files: string[];
}
