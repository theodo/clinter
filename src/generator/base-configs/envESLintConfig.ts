import { Linter } from "eslint";

export const browserESLintEnvConfig: Linter.Config = {
  env: {
    browser: true,
    es6: true,
  },
};

export const nodeESLintEnvConfig: Linter.Config = {
  env: {
    node: true,
    es6: true,
  },
};

export const jestESLintEnvConfig: Linter.Config = {
  env: {
    jest: true,
  },
};

export const jestESLintDependencies = ["eslint-plugin-jest"];
