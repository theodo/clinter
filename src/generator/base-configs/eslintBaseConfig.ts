import { Linter } from "eslint";

export const eslintBaseConfig: Linter.Config = {
  root: true,
  env: {},
  parserOptions: {},
  extends: ["eslint:recommended"],
  overrides: [],
  plugins: ["import"],
  rules: {
    curly: ["error", "all"],
    eqeqeq: ["error", "smart"],
    complexity: ["error", 8],
    "import/no-extraneous-dependencies": [
      "error",
      { devDependencies: true, optionalDependencies: false, peerDependencies: false },
    ],
    "no-shadow": [
      "error",
      {
        hoist: "all",
      },
    ],
    "prefer-const": "error",
    "import/order": [
      "error",
      {
        groups: [["external", "builtin"], "internal", ["parent", "sibling", "index"]],
      },
    ],
    "sort-imports": [
      "error",
      {
        ignoreCase: true,
        ignoreDeclarationSort: true,
        ignoreMemberSort: false,
        memberSyntaxSortOrder: ["none", "all", "multiple", "single"],
      },
    ],
  },
};

export const eslintBaseDependencies = ["eslint", "eslint-plugin-import"];
