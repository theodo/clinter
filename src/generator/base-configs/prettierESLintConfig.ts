import { Linter } from "eslint";

export const prettierESLintConfig: Linter.Config = {
  extends: ["plugin:prettier/recommended"],
};

export const prettierESLintDependencies = ["prettier", "eslint-config-prettier", "eslint-plugin-prettier"];
