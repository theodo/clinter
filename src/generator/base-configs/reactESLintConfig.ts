import { Linter } from "eslint";

export const reactESLintConfig: Linter.Config = {
  settings: {
    react: {
      version: "detect",
    },
  },
  extends: ["react-app", "plugin:jsx-a11y/recommended"],
  rules: {
    "react/no-string-refs": "warn",
    "react-hooks/rules-of-hooks": "error",
    "react-hooks/exhaustive-deps": "warn",
  },
  plugins: ["jsx-a11y"],
};

export const reactTypescriptESLintConfig: Linter.Config = {
  rules: {
    "@typescript-eslint/explicit-function-return-type": "off",
  },
};

export const reactESLintDependencies = ["eslint-plugin-react-app", "eslint-config-react-app", "eslint-plugin-jsx-a11y"];
