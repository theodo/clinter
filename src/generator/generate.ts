import { Linter } from "eslint";
import { generateTestESLintConfig, getTestESLintDependencies } from "generator/test-framework";
import { generateTypescriptESLintConfig, getTypescriptESLintDependencies } from "generator/typescript";
import { generateEnvESLintConfig } from "generator/env";
import { generateFrontFrameworkESLintConfig, getFrontFrameworkESLintDependencies } from "generator/front-framework";
import { generatePrettierESlintConfig, getPrettierESLintDependencies } from "generator/prettier";
import { eslintBaseConfig, eslintBaseDependencies } from "generator/base-configs/eslintBaseConfig";
import { ESLintDependencyGenerator, ESLintGenerator } from "generator/types";
import { mergeArrays, pipe } from "utils/utility";
import { ProjectInfoObject } from "types";

function addEslintOverride(
  overrides: Linter.ConfigOverride[],
  newOverride: Linter.ConfigOverride
): Linter.ConfigOverride[] {
  if (overrides.length === 0) {
    return [newOverride];
  }

  const [currentOverride, ...rest] = overrides;

  if (currentOverride.parser === newOverride.parser) {
    return [concatConfig(newOverride)(currentOverride), ...rest];
  }

  return [currentOverride, ...addEslintOverride(rest, newOverride)];
}

function concatEslintOverrides(
  prevOverrides: Linter.ConfigOverride[],
  nextOverrides: Linter.ConfigOverride[]
): Linter.ConfigOverride[] {
  return nextOverrides.reduce(addEslintOverride, prevOverrides);
}

function concatESlintObjects<T extends Record<string, unknown>>(prevObject: T, nextObject: T): T {
  return { ...prevObject, ...nextObject };
}

function concatESlintArrays<T extends Array<string> | string>(prev: T, next: T): Array<string> {
  const prevArray = (typeof prev === "string" ? [prev] : prev) as string[];
  const nextArray = (typeof next === "string" ? [next] : next) as string[];

  return mergeArrays(prevArray, nextArray);
}

export function concatConfig<T extends Linter.Config | Linter.ConfigOverride>(config: T): (prevConfig: T) => T {
  return (prevConfig) => ({
    ...prevConfig,
    ...config,
    extends: concatESlintArrays(prevConfig.extends ?? [], config.extends ?? []),
    rules: concatESlintObjects(prevConfig.rules ?? {}, config.rules ?? {}),
    plugins: concatESlintArrays(prevConfig.plugins ?? [], config.plugins ?? []),
    parserOptions: concatESlintObjects(prevConfig.parserOptions ?? {}, config.parserOptions ?? {}),
    env: concatESlintObjects(prevConfig.env ?? {}, config.env ?? {}),
    overrides: concatEslintOverrides(prevConfig.overrides ?? [], config.overrides ?? []),
  });
}

export function concatDependencies(deps2: string[]): (deps2: string[]) => string[] {
  return (deps1) => mergeArrays(deps1, deps2);
}

const generateBaseESLintConfig: ESLintGenerator = (_userAnswers) => {
  return concatConfig(eslintBaseConfig);
};

const getBaseESLintDependencies: ESLintDependencyGenerator = (_userAnswers) =>
  concatDependencies(eslintBaseDependencies);

export function generateEslintConfig(userAnswers: ProjectInfoObject, startConfig: Linter.Config = {}): Linter.Config {
  return pipe(
    generateBaseESLintConfig(userAnswers),
    generateTypescriptESLintConfig(userAnswers),
    generateEnvESLintConfig(userAnswers),
    generateTestESLintConfig(userAnswers),
    generateFrontFrameworkESLintConfig(userAnswers),
    // Prettier must be at the end of the list to avoid potential conflicts
    generatePrettierESlintConfig(userAnswers)
  )(startConfig);
}

export function getConfigDependencies(userAnswers: ProjectInfoObject): string[] {
  return pipe(
    getBaseESLintDependencies(userAnswers),
    getTypescriptESLintDependencies(userAnswers),
    getTestESLintDependencies(userAnswers),
    getFrontFrameworkESLintDependencies(userAnswers),
    getPrettierESLintDependencies(userAnswers)
  )([]);
}
