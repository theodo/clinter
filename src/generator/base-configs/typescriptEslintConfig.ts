import { Linter } from "eslint";
import { ESLintOverrider } from "generator/types";

export const typescriptBaseEslintConfig: Linter.Config = {
  extends: ["plugin:@typescript-eslint/recommended"],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: "tsconfig.json",
  },
  rules: {
    // Prefer using concise optional chain expressions instead of chained logical ands
    "@typescript-eslint/prefer-optional-chain": "error",
    // Disable JS no-shadow rule
    "no-shadow": "off",
    // Disallow variable declarations from shadowing variables declared in the outer scope
    "@typescript-eslint/no-shadow": "error",
  },
};

export const typescriptTypeEslintConfig: Linter.Config = {
  extends: ["plugin:@typescript-eslint/recommended-requiring-type-checking"],
  rules: {
    // Enforce the usage of the nullish coalescing operator instead of logical chaining
    "@typescript-eslint/prefer-nullish-coalescing": "error",
    // Restricts the types allowed in boolean expressions
    "@typescript-eslint/strict-boolean-expressions": [
      "error",
      {
        allowString: false,
        allowNumber: false,
        allowNullableObject: true,
      },
    ],
    // Flags unnecessary equality comparisons against boolean literals
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error",
    // Prevents conditionals where the type is always truthy or always falsy
    "@typescript-eslint/no-unnecessary-condition": "error",
    // Enforces that type arguments will not be used if not required
    "@typescript-eslint/no-unnecessary-type-arguments": "error",
    // Enforce the use of String#startsWith and String#endsWith instead of other equivalent methods of checking substrings
    "@typescript-eslint/prefer-string-starts-ends-with": "error",
    // Exhaustiveness checking in switch with union type
    "@typescript-eslint/switch-exhaustiveness-check": "error",
    // Prevents the text 'undefined' to appear in a template string
    "@typescript-eslint/restrict-template-expressions": [
      "error",
      {
        allowNumber: true,
        allowBoolean: true,
      },
    ],
  },
};

export const tsOverrider: ESLintOverrider = {
  files: ["**/*.ts?(x)"],
};

export const typescriptESLintDependencies = ["@typescript-eslint/parser", "@typescript-eslint/eslint-plugin"];
