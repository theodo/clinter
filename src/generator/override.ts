import { Linter } from "eslint";
import { ESLintOverrider } from "generator/types";

/**
 * Transform a config into an override config
 * @param overrider the eslint attributes to be added to transform the config to an override config
 */
export function wrapConfigInOverride(overrider: ESLintOverrider): (config: Linter.Config) => Linter.Config {
  return (config) => ({
    overrides: [
      {
        ...overrider,
        ...config,
      },
    ],
  });
}
