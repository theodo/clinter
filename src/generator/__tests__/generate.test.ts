import { generateEslintConfig } from "generator/generate";
import { FormatterInfo, FrontFrameworkInfo, TestFrameworkInfo, TypescriptInfo } from "types";
import { eslintBaseConfig } from "generator/base-configs";

describe("generateEslintConfig", () => {
  describe("Generator Mode", () => {
    describe("Simple eslint config", () => {
      it("should return a simple eslint configuration when no special answers are provided", () => {
        expect(
          generateEslintConfig({
            typescript: TypescriptInfo.None,
            env: [],
            formatter: FormatterInfo.None,
            frontFramework: FrontFrameworkInfo.None,
            test: TestFrameworkInfo.None,
          })
        ).toStrictEqual(eslintBaseConfig);
      });
    });

    describe("TypeScript", () => {
      it("should return a TS configuration with type checking", () => {
        const config = generateEslintConfig({
          typescript: TypescriptInfo.WithTypeChecking,
          env: [],
          formatter: FormatterInfo.None,
          frontFramework: FrontFrameworkInfo.None,
          test: TestFrameworkInfo.None,
        });

        expect(config).toMatchSnapshot();
        expect(config.overrides?.[0].extends).toContain("plugin:@typescript-eslint/recommended");
        expect(config.overrides?.[0].extends).toContain(
          "plugin:@typescript-eslint/recommended-requiring-type-checking"
        );
      });

      it("should return a TS configuration without type checking", () => {
        const config = generateEslintConfig({
          typescript: TypescriptInfo.NoTypeChecking,
          env: [],
          formatter: FormatterInfo.None,
          frontFramework: FrontFrameworkInfo.None,
          test: TestFrameworkInfo.None,
        });

        expect(config).toMatchSnapshot();
        expect(config.overrides?.[0].extends).toContain("plugin:@typescript-eslint/recommended");
      });
    });

    describe("React", () => {
      it("should return a react configuration", () => {
        const config = generateEslintConfig({
          typescript: TypescriptInfo.None,
          env: [],
          formatter: FormatterInfo.None,
          frontFramework: FrontFrameworkInfo.React,
          test: TestFrameworkInfo.None,
        });

        expect(config).toMatchSnapshot();
        expect(config.extends).toContain("react-app");
      });
    });

    describe("React", () => {
      it("should return a Vue configuration", () => {
        const config = generateEslintConfig({
          typescript: TypescriptInfo.None,
          env: [],
          formatter: FormatterInfo.None,
          frontFramework: FrontFrameworkInfo.Vue,
          test: TestFrameworkInfo.None,
        });

        expect(config).toMatchSnapshot();
        expect(config.extends).toContain("plugin:vue/essential");
      });
    });
  });
});
