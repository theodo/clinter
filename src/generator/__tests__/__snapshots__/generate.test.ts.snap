// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEslintConfig Generator Mode React should return a Vue configuration 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
    "plugin:vue/essential",
  ],
  "overrides": Array [],
  "parserOptions": Object {
    "parser": "babel-eslint",
  },
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-console": "off",
    "no-debugger": "off",
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;

exports[`generateEslintConfig Generator Mode React should return a Vue configuration with prettier formatter 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
    "plugin:vue/essential",
    "plugin:prettier/recommended",
    "prettier/vue",
  ],
  "overrides": Array [],
  "parserOptions": Object {
    "parser": "babel-eslint",
  },
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-console": "off",
    "no-debugger": "off",
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;

exports[`generateEslintConfig Generator Mode React should return a react configuration 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
    "react-app",
    "plugin:jsx-a11y/recommended",
  ],
  "overrides": Array [],
  "parserOptions": Object {},
  "plugins": Array [
    "import",
    "jsx-a11y",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "react-hooks/exhaustive-deps": "warn",
    "react-hooks/rules-of-hooks": "error",
    "react/no-string-refs": "warn",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
  "settings": Object {
    "react": Object {
      "version": "detect",
    },
  },
}
`;

exports[`generateEslintConfig Generator Mode React should return a react configuration with prettier formatter 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
    "react-app",
    "plugin:jsx-a11y/recommended",
    "plugin:prettier/recommended",
    "prettier/react",
  ],
  "overrides": Array [],
  "parserOptions": Object {},
  "plugins": Array [
    "import",
    "jsx-a11y",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "react-hooks/exhaustive-deps": "warn",
    "react-hooks/rules-of-hooks": "error",
    "react/no-string-refs": "warn",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
  "settings": Object {
    "react": Object {
      "version": "detect",
    },
  },
}
`;

exports[`generateEslintConfig Generator Mode TypeScript should return a TS configuration with prettier formatter 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
    "plugin:prettier/recommended",
    "prettier/@typescript-eslint",
  ],
  "overrides": Array [
    Object {
      "env": Object {},
      "extends": Array [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      "files": Array [
        "**/*.ts?(x)",
      ],
      "overrides": Array [],
      "parser": "@typescript-eslint/parser",
      "parserOptions": Object {
        "project": "tsconfig.json",
      },
      "plugins": Array [],
      "rules": Object {
        "@typescript-eslint/prefer-nullish-coalescing": "error",
        "@typescript-eslint/prefer-optional-chain": "error",
        "@typescript-eslint/strict-boolean-expressions": "error",
      },
    },
  ],
  "parserOptions": Object {},
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;

exports[`generateEslintConfig Generator Mode TypeScript should return a TS configuration with type checking 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
  ],
  "overrides": Array [
    Object {
      "env": Object {},
      "extends": Array [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      "files": Array [
        "**/*.ts?(x)",
      ],
      "overrides": Array [],
      "parser": "@typescript-eslint/parser",
      "parserOptions": Object {
        "project": "tsconfig.json",
      },
      "plugins": Array [],
      "rules": Object {
        "@typescript-eslint/prefer-nullish-coalescing": "error",
        "@typescript-eslint/prefer-optional-chain": "error",
        "@typescript-eslint/strict-boolean-expressions": "error",
      },
    },
  ],
  "parserOptions": Object {},
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;

exports[`generateEslintConfig Generator Mode TypeScript should return a TS configuration without type checking 1`] = `
Object {
  "env": Object {},
  "extends": Array [
    "eslint:recommended",
  ],
  "overrides": Array [
    Object {
      "extends": Array [
        "plugin:@typescript-eslint/recommended",
      ],
      "files": Array [
        "**/*.ts?(x)",
      ],
      "parser": "@typescript-eslint/parser",
      "parserOptions": Object {
        "project": "tsconfig.json",
      },
      "rules": Object {
        "@typescript-eslint/prefer-optional-chain": "error",
      },
    },
  ],
  "parserOptions": Object {},
  "plugins": Array [
    "import",
  ],
  "root": true,
  "rules": Object {
    "complexity": Array [
      "error",
      8,
    ],
    "curly": Array [
      "error",
      "all",
    ],
    "eqeqeq": Array [
      "error",
      "smart",
    ],
    "import/no-extraneous-dependencies": Array [
      "error",
      Object {
        "devDependencies": true,
        "optionalDependencies": false,
        "peerDependencies": false,
      },
    ],
    "import/order": Array [
      "error",
      Object {
        "groups": Array [
          Array [
            "external",
            "builtin",
          ],
          "internal",
          Array [
            "parent",
            "sibling",
            "index",
          ],
        ],
      },
    ],
    "no-shadow": Array [
      "error",
      Object {
        "hoist": "all",
      },
    ],
    "prefer-const": "error",
    "sort-imports": Array [
      "error",
      Object {
        "ignoreCase": true,
        "ignoreDeclarationSort": true,
        "ignoreMemberSort": false,
        "memberSyntaxSortOrder": Array [
          "none",
          "all",
          "multiple",
          "single",
        ],
      },
    ],
  },
}
`;
